### ISSUE TITLE
Forget to put the global-step in lr scheduler in train.py

### ISSUE THREAD

**Issue by SuperCB** [huggingface/transformers#25079]
### System Info

- ` transformers` version: 4.30.2 
- Platform: Linux-3.10.0-1160.83.1.el7.x86_64-x86_64-with-glibc2.35
- Python version: 3.10.6  
- Huggingface_hub version: 0.15.1 
- Safetensors version: 0.3.1  
- PyTorch version (GPU?): 2.0.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed 
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@sgugger 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

We fine-tuned the llama model using trainer.py and accelerated the training process using DeepSpeed. We chose the cosine function as the type for the lr scheduler. 

### Expected behavior

In the process of training LLM using `train()` in trainer.py, we observed that after reloading the model from a checkpoint with deepspeed zero-3, the learning rate scheduler did not resume iteration from the previous training's learning rate. Instead, there were gaps in the learning rate progression.We used a lr scheduler of the cosine type, which is not provided by Deepspeed. Instead, we observed that the trainer used the [lr_scheduler](https://github.com/pytorch/pytorch/blob/main/torch/optim/lr_scheduler.py#L124) defined in torch.

**Comment by sgugger** [comment]
cc @pacman100

**Comment by pacman100** [comment]
Hello, a minimal reproducer for this please. The checkpointing test loads back the correct lr: https://github.com/huggingface/accelerate/actions/runs/5651936442/job/15310803049#step:5:2259

**Comment by Dounm** [comment]
> Hello, a minimal reproducer for this please. The checkpointing test loads back the correct lr: https://github.com/huggingface/accelerate/actions/runs/5651936442/job/15310803049#step:5:2259

Hello, the difference between your job and ours lies in the **different lr_scheduler type**. We use the cosine lr_scheculer (with `--lr_scheduler_type cosine`) instead of set `"scheduler"` in `deepspeed_config.json`.

What we see in wandb is, the `train/learning_rate` always restart from zero with warmups when we resume training from checkpoints.
![image](https://github.com/huggingface/transformers/assets/9065640/e928d3df-7194-456e-9652-cfd405b30aa3)


As we go through the code(correct us if we were wrong), the reason  is that DeepSpeed doesn't support consine lr_scheduler, so Transformers use [pytorch native lr_scheduler](https://github.com/pytorch/pytorch/blob/e18d53e2df44bccd7231cdf3dad6ea1255221bd4/torch/optim/lr_scheduler.py#L124), which maintains a self-incrementing variable `self.last_epoch` if we don't pass `epoch` when calling `step()`.

We will provide a minimal reproducer later.

**Comment by SuperCB** [comment]
We used [Stanford Alpaca](https://github.com/tatsu-lab/stanford_alpaca) to reproduce this bug in the [repo](https://github.com/SuperCB/fix_trainer), and you only need to execute run.sh to run it. Record the changes in the learning rate before and after reloading the checkpoint, and you can observe significant discontinuities."

**Comment by Dounm** [comment]
@pacman100 Hello, I apologize for bothering you. Could you please let us know if there have been any updates or if there might be any corrections needed in our description? Much thanks.

**Comment by pacman100** [comment]
Hello @Dounm, please see https://github.com/huggingface/transformers/issues/25865#issuecomment-1700722492 and https://github.com/huggingface/transformers/issues/25865#issuecomment-1702239905. This should be resolving this issue.
